src/control/server/ctl_system.go

```diff
@@ -33,81 +35,159 @@ import (
+			return
+		}
 	default:
-		return nil, errors.New("unrecognised system ranks method") // programming error
+		// empty rank/host sets implies include all ranks so pass empty
+		// string to CheckRanks()
+		if hitRS, missRS, err = svc.membership.CheckRanks(""); err != nil {
+			return
+		}
 	}
 
-	return fn(ctx, invoker, req)
+	if missHS == nil {
+		missHS = new(hostlist.HostSet)
+	}
+	if missRS == nil {
+		missRS = new(system.RankSet)
+	}
+
+	return
+}
+
+// unpackSysReq takes the protobuf request, selects on its type and returns the
+// client call object and force flag if set in request.
+func unpackSysReq(pbReq interface{}) (_ systemRanksFunc, _ bool, _ string, _ string, err error) {
+	if pbReq == nil {
+		return nil, false, "", "", errors.New("nil request")
+	}
+
+	switch v := pbReq.(type) {
+	case *ctlpb.SystemQueryReq:
+		if v == nil {
+			err = errors.Errorf("nil %T request", v)
+			return
+		}
+		return control.PingRanks, false, v.GetHosts(), v.GetRanks(), nil
+	case *ctlpb.SystemPrepShutdownReq:
+		fmt.Printf("processing prepare progress\n")
+		if v == nil {
+			err = errors.Errorf("nil %T request", v)
+			return
+		}
+		return control.PrepShutdownRanks, false, v.GetHosts(), v.GetRanks(), nil
+	case *ctlpb.SystemStopReq:
+		if v == nil {
+			err = errors.Errorf("nil %T request", v)
+			return
+		}
+		return control.StopRanks, v.GetForce(), v.GetHosts(), v.GetRanks(), nil
+	case *ctlpb.SystemResetFormatReq:
+		if v == nil {
+			err = errors.Errorf("nil %T request", v)
+			return
+		}
+		return control.ResetFormatRanks, false, v.GetHosts(), v.GetRanks(), nil
+	case *ctlpb.SystemStartReq:
+		if v == nil {
+			err = errors.Errorf("nil %T request", v)
+			return
+		}
+		return control.StartRanks, false, v.GetHosts(), v.GetRanks(), nil
+	default:
+		err = errors.Errorf("unknown request type: %T", v)
+		return
+	}
 }
 
 // rpcFanout sends requests to ranks in list on their respective host
 // addresses through functions implementing UnaryInvoker.
 //
-// The fan-out host list is derived from the provided rank list and contains
-// any host that is managing a rank selected by the request rank list filter.
+// Required client method and any force flag in request are returned from
+// processSysReq().
+//
+// The fan-out host and rank lists are resolved by calling resolveRanks().
+//
+// Pass true as last parameter to update member states on request failure.
 //
 // Fan-out is invoked by control API *Ranks functions.
-func (svc *ControlService) rpcFanout(ctx context.Context, req *control.RanksReq, method ranksMethod) (system.MemberResults, error) {
-	if svc.membership == nil {
-		return nil, errors.New("nil system membership")
+func (svc *ControlService) rpcFanout(parent context.Context, pbReq interface{}, updateOnFail bool) (*fanoutResponse, *system.RankSet, error) {
+	// derive fanout call and extract force flag, hosts and ranks from pbReq
+	method, force, hosts, ranks, err := unpackSysReq(pbReq)
```

Comment:
the method can only be known through a type assertion on the request, that's why we have the switch in the unpack

Reply:
But wouldn't the callers of rpcFanout know what sort of request they have already, and thus be able to pass the method along? That's the point I was trying to get to. Then all we'd need to figure out at this level is the ranks/hosts, which we can get via interfaces instead of knowing the specific type.

Reply:
Initially I had this design where the caller passed in fanoutRequest struct with the info as you describe but I decided it was better with the unpack method as I have. I will revert to that design so I can get this landed.